[Aqara] Roller shade controller

sbkimaqara opened the pull request:
A roller shade controller device was added. Codes commonly used with the curtain controllers are separated into separate files. Functions of each device are as follows.
- Common: Open, Close, Pause, Level, Reverse direction
- Roller shade controller: Tilt up, Tilt down, Guided initialized state
- Curtain controller: Initialized state, Soft touch

greens commented on drivers/SmartThings/zigbee-window-treatment/capabilities/initializedStateWithGuide.yaml:
```diff
@@ -0,0 +1,20 @@
+id: stse.initializedStateWithGuide
```
can't you just use your _other_ initialized state custom capability

we'd really prefer to add as few custom capabilities into _our_ repo as possible

Maybe, if this is installation instructions, it would be better to include in the catalog info on installation?

Kwang-Hui replied:
Hi @greens , it is related with plugin feature. let me give you more details on slack

greens replied:
Ideally, for tiny revisions to existing capabilities you have already published, the best course of action might just be to update those with any changes you need rather than creating another new, largely identical, capability. This is, as you can see, leading to a drastic increase in your code complexity.

greens commented on drivers/SmartThings/zigbee-window-treatment/profiles/aqara-curtain.yml:
```diff
@@ -1,4 +1,4 @@
-name: window-treatment-aqara
+name: aqara-curtain
```
once a profile has been created, it cannot be moved/renamed as this can create "orphaned" devices that are unaware of the change

sbkimaqara replied:
Okay. I will change the name as below.
- Curtain controller: window-treatment-aqara.yml
- Roller Shade Controller: window-treatment-aqara-roller-shade-rotate.yml

greens commented on drivers/SmartThings/zigbee-window-treatment/capabilities/shadeRotateState.yaml:
```diff
@@ -0,0 +1,33 @@
+id: stse.shadeRotateState
+version: 1
+status: proposed
+name: shade Rotate State
```
can you explain a bit more about what this is and how it's distinct from our other window shade devices that use the `reverse` preference?

Kwang-Hui replied:
Hi @greens , it's due to translation support as mentioned at slack

greens commented on drivers/SmartThings/zigbee-window-treatment/src/aqara/aqara_utils.lua:
```diff
@@ -0,0 +1,212 @@
+  send_lift_percentage_cmd(device, command, level)
+end
+
+local function shade_open_cmd(driver, device, command)
+  send_lift_percentage_cmd(device, command, 100)
+end
+
+local function shade_close_cmd(driver, device, command)
+  send_lift_percentage_cmd(device, command, 0)
+end
```
both of these are duplicated above

sbkimaqara replied:
I'll edit and push.

greens commented on drivers/SmartThings/zigbee-window-treatment/src/aqara/aqara_utils.lua:
```diff
@@ -0,0 +1,212 @@
+  end
+end
+
+local function emit_shade_level_event(device, level)
+  device:emit_event(capabilities.windowShadeLevel.shadeLevel(level))
+end
```
Why are there so many one-line functions? I don't think this significantly increases readability.

sbkimaqara replied:
I'll edit and push.

greens commented on drivers/SmartThings/zigbee-window-treatment/src/aqara/aqara_utils.lua:
```diff
@@ -0,0 +1,212 @@
+
+local aqara_utils = {}
+
+local function read_private_attribute(device, cluster_id, attribute_id)
+  local message = cluster_base.read_attribute(device, data_types.ClusterId(cluster_id), attribute_id)
+  message.body.zcl_header.frame_ctrl:set_mfg_specific()
+  message.body.zcl_header.mfg_code = data_types.validate_or_build_type(MFG_CODE, data_types.Uint16, "mfg_code")
+  device:send(message)
+end
```
we provide this with `cluster_base.read_manufacturer_specific_attribute` in `local cluster_base = require "st.zigbee.cluster_base"`

There are examples in other drivers.

sbkimaqara replied:
I'll edit and push.

greens commented on drivers/SmartThings/zigbee-window-treatment/src/aqara/init.lua:
```diff
@@ -1,301 +1,21 @@
-    [capabilities.windowShadeLevel.ID] = {
-      [capabilities.windowShadeLevel.commands.setShadeLevel.NAME] = window_shade_level_cmd
-    },
-    [capabilities.windowShade.ID] = {
-      [capabilities.windowShade.commands.open.NAME] = window_shade_open_cmd,
-      [capabilities.windowShade.commands.close.NAME] = window_shade_close_cmd,
-      [capabilities.windowShade.commands.pause.NAME] = window_shade_pause_cmd
-    },
```
as far as I can tell, both sub-drivers _at least_ still have the same open, close, and pause commands. They can inherit them from the parent driver rather than making this driver completely empty.

greens replied:
`refresh` is also the same for both

greens replied:
`shade_level_report_handler` also seems to be the same for both.

greens replied:
You should re-examine what functions of these two sub-drivers are _actually_ distinct. Everything in common should be here in the base driver.

sbkimaqara replied:
Where can I find a example code to inherit sub-driver from the parent?

greens replied:
Include `sub_drivers = {...}` in your sub-driver. Any handler in the sub-driver that isn't overwritten by the child will default to the parent's.

see, for example, the `axis` driver: https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/tree/main/drivers/SmartThings/zigbee-window-treatment/src/axis

or your own motion sensor driver:https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/tree/main/drivers/SmartThings/zigbee-motion-sensor/src/aqara

sbkimaqara replied:
I updated the code to override the child's handler according to your guide.

greens commented on drivers/SmartThings/zigbee-window-treatment/src/aqara/aqara_utils.lua:
```diff
@@ -0,0 +1,212 @@
+
+local function is_matched_profile(device, profile)
+  for _, fingerprint in pairs(FINGERPRINTS) do
+    if device:get_model() == fingerprint.model and fingerprint.device_profile == profile then
+      return true
+    end
```
This function is confusing for me. What's the advantage of the "profile" field over just matching on the model?

sbkimaqara replied:
I'll edit and push.

greens commented on drivers/SmartThings/zigbee-window-treatment/src/aqara/aqara_utils.lua:
```diff
@@ -0,0 +1,212 @@
+  write_private_attribute(device, Basic.ID, PREF_ATTRIBUTE_ID, data_types.CharString, str)
+end
+
+local function write_reverse_pref_default(device, str)
+  write_pref_attribute(device, PREF_REVERSE_DEFAULT)
+end
```
unused

sbkimaqara replied:
I'll edit and push.

greens commented on drivers/SmartThings/zigbee-window-treatment/src/aqara/curtain/init.lua:
```diff
@@ -0,0 +1,234 @@
+  aqara_utils.write_pref_attribute(device, PREF_INITIALIZE)
+end
+
+local function setInitializationField(device, value)
+  device:set_field(INIT_STATE, value)
+end
```
this is only used in the next function and then you go back to calling `device:set_field`

I recommend using `device:set_field` everywhere

sbkimaqara replied:
I'll edit and push.

sbkimaqara commented:
I've updated everything to the code except for one of your advice. If you tell me how to improve the code with an inheritance structure, I will edit that too. 😀

greens commented on drivers/SmartThings/zigbee-window-treatment/src/aqara/roller-shade/init.lua:
```diff
@@ -0,0 +1,159 @@
+
+local function window_shade_open_cmd(driver, device, command)
+  -- Cannot be controlled if not initialized
+  local initialized = device:get_field(INITIALIZED_STATE) or 0
```
you should be able to use `device:get_latest_state` for your custom capability rather than have this separate field

greens replied:
e.g.:

```
local function window_shade_open_cmd(driver, device, command)
  local initialized = device:get_latest_state("main", initializedStateWithGuide.ID, initializedStateWithGuide.initializedStateWithGuide.NAME) or 0
  if initialized == initializedStateWithGuide.initializedStateWithGuide.initialized.NAME then
    device:send_to_component(...)
  end
end
```

something like that

having only one exit point for the code is easier to read as well

sbkimaqara replied:
Thanks. I'll update it.

greens commented on drivers/SmartThings/zigbee-window-treatment/src/aqara/init.lua:
```diff
@@ -47,218 +43,170 @@ local is_aqara_products = function(opts, driver, device)
-local do_refresh = function(self, device)
-  device:send(AnalogOutput.attributes.PresentValue:read(device))
+  -- initialize
+  device:set_field(INIT_STATE, INIT_STATE_INIT)
```
same `get_latest_state` can be used here

greens commented on drivers/SmartThings/zigbee-window-treatment/src/aqara/version/init.lua:
```diff
@@ -0,0 +1,77 @@
+
+  -- initializedState
+  local state = value.value
+  if state == aqara_utils.SHADE_STATE_STOP or state == 0x04 then
```
what's `4`? is this `4` the only different with the base handler?

greens replied:
if this is the only difference, and this device is the only one that reports `4`, it can just be included in the base driver and you don't have to override the handler here. Does that makes sense? If the non-"version" device never send the value, it doesn't matter if this check is in the base handler or not

sbkimaqara commented:
@greens 
4 is the only different with the base handler.

sbkimaqara commented:
@greens 
I agree with you, so I've updated the following.
- A handler merged into parent code.
- Re-named function name.
- Removed unused code.

greens commented on drivers/SmartThings/zigbee-window-treatment/src/aqara/version/init.lua:
```diff
@@ -0,0 +1,37 @@
+  aqara_utils.emit_shade_event(device, value)
+end
+
+local function shade_level_report_handler(driver, device, value, zb_rx)
+  aqara_utils.emit_shade_level_event(device, value)
+  aqara_utils.emit_shade_event(device, value)
+end
```
this is still identical to the parent driver and doesn't need to be overwritten

greens commented on drivers/SmartThings/zigbee-window-treatment/src/aqara/version/init.lua:
```diff
@@ -0,0 +1,37 @@
+  zigbee_handlers = {
+    attr = {
+      [WindowCovering.ID] = {
+        [WindowCovering.attributes.CurrentPositionLiftPercentage.ID] = shade_level_report_legacy_handler
```
based on how this is written, it might make sense to make the base driver handle this attribute and then invert the `can_handle` of this driver to be `version ~= 34` and only in that case ignore the value

greens commented on drivers/SmartThings/zigbee-window-treatment/src/aqara/roller-shade/init.lua:
```diff
@@ -0,0 +1,150 @@
+end
+
+local function write_tilt_attribute(device, payload)
+  local value = data_types.validate_or_build_type(payload, data_types.Uint16, "payload")
+  local message = cluster_base.write_attribute(device, data_types.ClusterId(MULTISTATE_CLUSTER_ID),
+    data_types.AttributeId(MULTISTATE_ATTRIBUTE_ID), value)
+  local frm_ctrl = FrameCtrl(0x10)
+  message.body.zcl_header.frame_ctrl:set_mfg_specific()
+  message.body.zcl_header.mfg_code = data_types.validate_or_build_type(aqara_utils.MFG_CODE, data_types.Uint16,
+    "mfg_code")
+  message.body.zcl_header.frame_ctrl = frm_ctrl
```
all of this is equivalent to `cluster_base.write_manufacturer_specific_attribute`

greens replied:
also you've immediately overwritten the result from calling `set_mfg_specific` with your own `FrameCtrl` object immediately afterwards

greens commented on drivers/SmartThings/zigbee-window-treatment/src/aqara/init.lua:
```diff
@@ -47,218 +40,160 @@ local is_aqara_products = function(opts, driver, device)
 
-local function send_open_cmd(device, component)
-  device:send_to_component(component, WindowCovering.server.commands.GoToLiftPercentage(device, 100))
+local function window_shade_pause_cmd(driver, device, command)
+  device:send_to_component(command.component, WindowCovering.server.commands.Stop(device))
 end
```
this is the same as the default base driver and doesn't need to be overwritten.

sbkimaqara commented:
> this is still identical to the parent driver and doesn't need to be overwritten

I deleted the code.

> based on how this is written, it might make sense to make the base driver handle this attribute and then invert the can_handle of this driver to be version ~= 34 and only in that case ignore the value

I updated the code.

> all of this is equivalent to cluster_base.write_manufacturer_specific_attribute

I updated the code.

> also you've immediately overwritten the result from calling set_mfg_specific with your own FrameCtrl object immediately afterwards

I updated the code.

> this is the same as the default base driver and doesn't need to be overwritten.

I deleted the code. I also deleted the open/close command handlers because they are same to the default base driver.

greens commented on drivers/SmartThings/zigbee-window-treatment/src/aqara/roller-shade/init.lua:
```diff
@@ -0,0 +1,140 @@
+    if state == "rotateUp" then
+      local message = cluster_base.write_manufacturer_specific_attribute(device, MULTISTATE_CLUSTER_ID,
+        MULTISTATE_ATTRIBUTE_ID, aqara_utils.MFG_CODE, data_types.Uint16, ROTATE_UP_VALUE)
+      message.body.zcl_header.frame_ctrl = FrameCtrl(0x10)
```
The header value is a bitmask, right? `0x10` means disable default response, but in doing this you've unset the mfg-specific bit. I'd assume you'd want both of them enabled, right? In that case you'd do:

`message.body.zcl_header.frame_ctrl:set_disable_default_response()`

sbkimaqara commented:
@greens 
I apologize for my mistake. I pressed resolve conversation button. 

> @greens commented on this pull request.
> 
> In [drivers/SmartThings/zigbee-window-treatment/src/aqara/roller-shade/init.lua](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/pull/423#discussion_r1089350277):
> 
> +    device:send_to_component(command.component, WindowCovering.server.commands.GoToLiftPercentage(device, 0))
> +  end
> +end
> +
> +local function set_rotate_command_handler(driver, device, command)
> +  device:emit_event(shadeRotateState.rotateState.idle()) -- update UI
> +
> +  -- Cannot be controlled if not initialized
> +  local initialized = device:get_latest_state("main", initializedStateWithGuide.ID,
> +    [initializedStateWithGuide.initializedStateWithGuide.NAME](http://initializedstatewithguide.initializedstatewithguide.name/)) or 0
> +  if initialized == [initializedStateWithGuide.initializedStateWithGuide.initialized.NAME](http://initializedstatewithguide.initializedstatewithguide.initialized.name/) then
> +    local state = command.args.state
> +    if state == "rotateUp" then
> +      local message = cluster_base.write_manufacturer_specific_attribute(device, MULTISTATE_CLUSTER_ID,
> +        MULTISTATE_ATTRIBUTE_ID, aqara_utils.MFG_CODE, data_types.Uint16, ROTATE_UP_VALUE)
> +      message.body.zcl_header.frame_ctrl = FrameCtrl(0x10)
> The header value is a bitmask, right? 0x10 means disable default response, but in doing this you've unset the mfg-specific bit. I'd assume you'd want both of them enabled, right? In that case you'd do:
> 
> message.body.zcl_header.frame_ctrl:set_disable_default_response()
>

sbkimaqara commented:
@greens 

To control the device's rotate-up, the frame_ctrl value must be set to 0x10. This is a device requirement. If the message was sent with "message.body.zcl_header.frame_ctrl:set_disable_default_response()", the frame_ctrl value is set to 0x14. In this case, the rotate-up does not work.
I attached the log below.

Case 0x10:

> < ZigbeeMessageTx || Uint16: 0x0000, < AddressHeader || src_addr: 0x0000, src_endpoint: 0x01, dest_addr: 0x2466, dest_endpoint: 0x01, profile: 0x0104, cluster: 0x0013 >, < ZCLMessageBody || < ZCLHeader || ZclFrameCtrl: 0x10, seqno: 0x00, ZCLCommandId: 0x02 >, < WriteAttribute || < AttributeRecord || attr_id: 0x0055, DataType: Uint16, data: 0x0004 > > > >
> < ZigbeeMessageRx || type: 0x00, < AddressHeader || src_addr: 0x2466, src_endpoint: 0x01, dest_addr: 0x0000, dest_endpoint: 0x01, profile: 0x0104, cluster: 0x0013 >, lqi: 0xC8, rssi: -50, body_length: 0x0004, < ZCLMessageBody || < ZCLHeader || frame_ctrl: 0x18, seqno: 0x18, ZCLCommandId: 0x04 >, < WriteAttributeReponse || ZclStatus: SUCCESS > > >

Case set_disable_default_response():

> < ZigbeeMessageTx || Uint16: 0x0000, < AddressHeader || src_addr: 0x0000, src_endpoint: 0x01, dest_addr: 0x2466, dest_endpoint: 0x01, profile: 0x0104, cluster: 0x0013 >, < ZCLMessageBody || < ZCLHeader || frame_ctrl: 0x14, mfg_code: 0x115F, seqno: 0x00, ZCLCommandId: 0x02 >, < WriteAttribute || < AttributeRecord || attr_id: 0x0055, DataType: Uint16, data: 0x0004 > > > >
> < ZigbeeMessageRx || type: 0x00, < AddressHeader || src_addr: 0x2466, src_endpoint: 0x01, dest_addr: 0x0000, dest_endpoint: 0x01, profile: 0x0104, cluster: 0x0013 >, lqi: 0xC4, rssi: -40, body_length: 0x0008, < ZCLMessageBody || < ZCLHeader || frame_ctrl: 0x1C, mfg_code: 0x115F, seqno: 0x13, ZCLCommandId: 0x04 >, < WriteAttributeReponse || < WriteAttributeResponseRecord || ZclStatus: UNSUPPORTED_ATTRIBUTE, AttributeId: 0x0055 > > > >

Thank you.

sbkimaqara commented:
@greens 
Would you tell me your plans for this PR to be merged into the main branch?

Kwang-Hui commented:
> @greens Would you tell me your plans for this PR to be merged into the main branch?

The certification test isn't completed yet. Once it has completed we can go for the next steps. Thanks

Kwang-Hui commented on drivers/SmartThings/zigbee-window-treatment/src/aqara/roller-shade/init.lua:
```diff
@@ -0,0 +1,140 @@
+end
+
+local function set_rotate_command_handler(driver, device, command)
+  device:emit_event(shadeRotateState.rotateState.idle()) -- update UI
```
```suggestion
  device:emit_event(shadeRotateState.rotateState.idle({state_change = true})) -- update UI
```
Hi @sbkimaqara , Could you add `state_change = true` option to avoid network error on latest plugin?

lelandblue reviewed (commented):
@greens Can you please review these changes?

lelandblue reviewed (commented):
Hello @Kwang-Hui We received this request
